mypy/checker.py

```diff
@@ -265,119 +265,125 @@ def check_overlapping_overloads(self, defn: OverloadedFuncDef) -> None:
-    # - get_generator_receive_type(t, c) returns ts.
+    # - get_generator_receive_type(t, c) returns tc.
     # - get_generator_return_type(t, c) returns tr.
 
     def is_generator_return_type(self, typ: Type, is_coroutine: bool) -> bool:
         """Is `typ` a valid type for a generator/coroutine?
 
-        True if either Generator or Awaitable is a supertype of `typ`.
+        True if `typ` is a *supertype* of Generator or Awaitable.
+        Also true it it's *exactly* AwaitableGenerator (modulo type parameters).
         """
         if is_coroutine:
+            # This means we're in Python 3.5 or later.
             at = self.named_generic_type('typing.Awaitable', [AnyType()])
-            return is_subtype(at, typ)
+            if is_subtype(at, typ):
+                return True
         else:
             gt = self.named_generic_type('typing.Generator', [AnyType(), AnyType(), AnyType()])
-            return is_subtype(gt, typ)
+            if is_subtype(gt, typ):
+                return True
+        # Check that AwaitableGenerator exists -- it isn't defined in Python 2:
+        try:
+            self.lookup_qualified('typing.AwaitableGenerator')
+        except KeyError:
+            return False
+        agt = self.named_generic_type('typing.AwaitableGenerator',
+                                      [AnyType(), AnyType(), AnyType(), AnyType()])
+        return is_equivalent(typ, agt)
 
     def get_generator_yield_type(self, return_type: Type, is_coroutine: bool) -> Type:
         """Given the declared return type of a generator (t), return the type it yields (ty)."""
         if isinstance(return_type, AnyType):
             return AnyType()
         elif not self.is_generator_return_type(return_type, is_coroutine):
-            # If the function doesn't have a proper Generator (or superclass) return type, anything
-            # is permissible.
+            # If the function doesn't have a proper Generator (or
+            # Awaitable) return type, anything is permissible.
             return AnyType()
         elif not isinstance(return_type, Instance):
             # Same as above, but written as a separate branch so the typechecker can understand.
             return AnyType()
         elif return_type.type.fullname() == 'typing.Awaitable':
+            # Awaitable: ty is Any.
             return AnyType()
         elif return_type.args:
+            # AwaitableGenerator, Generator, Iterator, or Iterable; ty is args[0].
             ret_type = return_type.args[0]
             # TODO not best fix, better have dedicated yield token
             if isinstance(ret_type, NoneTyp):
                 return Void()
             return ret_type
         else:
-            # If the function's declared supertype of Generator has no type
-            # parameters (i.e. is `object`), then the yielded values can't
-            # be accessed so any type is acceptable.
+            # object: ty is Any.
```

Comment:
I think it's worth keeping the note that this is `Any` because the yielded values can't be accessed anywhere.